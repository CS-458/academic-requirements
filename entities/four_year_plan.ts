import type { TargetType } from "dnd-core";

// Defines the major object we are passing between files
export interface MajorType {
  id: number; // database major id
  name: string; // name of the major
}

// Defines the concentration object we are passing between files
export interface ConcentrationType {
  idConcentration: number; // Database concentration id
  name: string; // concentration name
  fourYearPlan: string; // this is a json of a fourYearPlan stored in a string
}

// Defines the course object as it is used in most locations
export interface CourseType {
  credits: number; // number of credits a course has
  name: string; // course name
  number: string; // course number STORED IN DATABASE AS STRING
  semesters: string; // whether it runs in fall or spring (FA or SP) or null (runs both)
  subject: string; // course subject i.e. CS, MATH
  preReq: string; // string containing prerequisite courses in a special format
  category: string; // the category the course falls into
  idCourse: number; // Database course id
  idCategory: number; // Database id for the category
  dragSource: string; // Which element it came from i.e. "Semester 1" or CourseList
  repeatableForCred: boolean; // If a course can be taken twice for credit
}

// Defines the actual draggable course object type
export interface DragCourseType {
  name: string, // see CourseType
  subject: string, // see CourseType
  number: string, // see CourseType
  type: string, // Don't worry about this, all assigned to course
  credits: number, // see CourseType
  semesters: string, // see CourseType
  preReq: string, // see CourseType
  dragSource: string, // see CourseType
  warningYellowColor: number | undefined, // If the course is in multiple semesters
  warningOrangeColor: number | undefined, // If the course is in the wrong semester fall vs spring
  warningRedColor: number | undefined, // If the course has a prereq error
  idCourse: number, // see CourseType
  idCategory: number // see CourseType
  repeatableForCred: boolean; // see CourseType
}

// Defines the props for the course list component
export interface CourseListType {
  accept: TargetType; // makes it a drop target
  onDrop: (item: any) => void; // function for what to do on drop
  courses: CourseType[]; // list of courses it displays for dragging
}

// Defines a requirement as it comes from the database
export interface RequirementType {
  courseCount: number; // number of courses required to satisfy
  courseReqs: string; // list of courses required to satisfy
  creditCount: number; // number of credits required to satisfy
  idCategory: number; // Database id for requirement
  name: string; // name of the requirement
  parentCategory: number; // Database id for parent requirement category
}

// Defines the saved information for each requirement
export interface RequirementComponentType {
  courseCount: number | null; // number of courses required to satisfy
  courseReqs: string | null; // list of courses required to satisfy
  creditCount: number; // number of credits required to satisfy
  idCategory: number; // Database id for requirement
  name: string; // name of the requirement
  shortName: string; // abbreviated version of the name for display
  parentCategory: number | null; // Database id for parent requirement category
  percentage: number; // Percentage that requirement has been filled
  inheritedCredits: number; // Number of credits a parent req inherits from child req
  coursesTaken: string; // list of courses already used toward requirement
  courseCountTaken: number; // number of courses taken towards requirement
  creditCountTaken: number; // number of credits taken towards requirement
}

// Defines a list of Requirements
export interface RequirementsType {
  [category: number]: RequirementType;
}

// Defines a list of courses in more than one category for requirements
export interface MultipleCategoriesType {
  categories: number[]; // list of category IDs the course falls under
  idString: string; // an identifier for the course i.e. CS-144
}

// Defines the properties of a semester component
export interface SemesterProps {
  accept: TargetType; // makes it a drop target, don't worry about it
  onDrop: (item: any) => void; // function for when an item is dropped
  semesterNumber: number; // the number of the semester
  courses: CourseType[]; // a list of courses in the semester
  SemesterCredits: number; // total credits of all courses in semester
  Warning: string; // credit warning (low or high)
  warningPrerequisiteCourses: CourseType[]; // list of courses in with prereq issues
  warningFallvsSpringCourses: CourseType[]; // list of courses with semester issues
  warningDuplicateCourses: CourseType[]; // list of courses in more than one semester
}

// Defines the properties of a semester object
export interface SemesterType {
  accepts: string[], // this is just a course constant
  semesterNumber: number, // number of the semester
  courses: CourseType[], // list of courses in semester
  SemesterCredits: number, // number of credits in the semester
  Warning: string // credit warning (high or low)
}

//  Defines the properties that should be passed in to the fourYearPlan page
export interface FourYearPlanType {
  PassedCourseList: CourseType[];
  requirements: RequirementComponentType[] | null | undefined;
  requirementsGen: RequirementComponentType[] | null | undefined;
  importData?: {};
}

export interface UserSavedSchedule {
  userID: string,
  name: string,
  timestamp: number,
  scheduleData: {
    Major: string,
    Concentration: string,
    "Completed Courses": string[],
    ClassPlan: {
      Semester1: string[],
      Semester2: string[],
      Semester3: string[],
      Semester4: string[],
      Semester5: string[],
      Semester6: string[],
      Semester7: string[],
      Semester8: string[]
    }
  }
}
