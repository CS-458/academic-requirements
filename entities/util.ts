let id = 0;
export function uniqueId(): string {
  return `id_${id++}`;
}
