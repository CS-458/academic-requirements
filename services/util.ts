/// Fetch a request (similar to fetch), but throw an error if the request fails
export async function fetchApi<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, init);
  if (!res.ok) {
    throw new Error(`Response code ${res.status} for request to ${url}`);
  }
  return await res.json();
}
