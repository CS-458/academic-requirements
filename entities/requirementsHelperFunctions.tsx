import { RequirementComponentType, CourseType } from "./four_year_plan";

// Moves child requirements into parent categories where appropriate
export function processRequirementLists(
  requirements: RequirementComponentType[],
  requirementsGen: RequirementComponentType[]):
  { req: RequirementComponentType[], reqGen: RequirementComponentType[], display: RequirementComponentType[] } {
  const temp: RequirementComponentType[] = [];
  let tempReqList: RequirementComponentType[] = requirements;
  const tempGen: RequirementComponentType[] = requirementsGen;
  tempReqList.forEach((x) => {
    if (x.parentCategory === null && !(x.courseReqs === null && x.creditCount === null && x.courseCount === null)) {
      temp.push(x);
    } else {
      tempReqList.forEach((n) => {
        if (n.idCategory === x.parentCategory) {
          if (n.courseReqs === null && n.creditCount === null && n.courseCount === null) {
            temp.push(x);
          }
        }
      });
    }

    if (x.parentCategory !== null) {
      // recursively moves gen ed requirements to their parent req
      function moveReqs(x: RequirementComponentType): void {
        let parent: RequirementComponentType | null = null;
        for (let i = 0; i < tempGen.length; i++) {
          // locate the parent requirement
          if (tempGen[i].idCategory === x.parentCategory) {
            // add any child reqs to the parent's reqs
            parent = tempGen[i];
            tempGen[i].inheritedCredits = x.creditCount;
            if (tempGen[i].courseReqs === null) {
              tempGen[i].courseReqs = x.courseReqs;
            } else if (tempGen[i].courseReqs?.includes(x.courseReqs ?? "") !== undefined) {
              tempGen[i].courseReqs = tempGen[i].courseReqs + "," + x.courseReqs;
            }
            tempGen[i].inheritedCredits = x.creditCount;
            tempReqList = tempReqList.filter((item) => item.idCategory !== x.idCategory);
          }
        }
        // if the parent has a parent call the function again
        if (parent?.parentCategory !== null && parent !== null) {
          moveReqs(parent);
        }
      }
      moveReqs(x);
    }
  });
  tempReqList.forEach((req) => { req.courseCountTaken = 0; req.coursesTaken = ""; req.creditCountTaken = 0; req.percentage = 0; });
  tempGen.forEach((x) => {
    if (x.parentCategory === null || x.parentCategory === undefined) {
      temp.push(x);
    }
  });
  tempGen.forEach((req) => { req.courseCountTaken = 0; req.coursesTaken = ""; req.creditCountTaken = 0; req.percentage = 0; });

  return { req: tempReqList, reqGen: tempGen, display: temp };
}

// Creates a list of courses that can potentially fill multiple requirements categories
export function createMultipleCategoryList(PassedCourseList: CourseType[]): { idString: string, categories: number[] } [] {
  const tempArr: {
    idString: string;
    categories: number[];
  }[] = [];
  // go through each item in the array to get any with duplicate categories
  for (let i = 0; i < PassedCourseList.length; i++) {
    let skip = false;
    // check that we haven't already added this on to the array
    for (let k = 0; k < tempArr.length; k++) {
      if (
        PassedCourseList[i].subject + "-" + PassedCourseList[i].number ===
        tempArr[k].idString
      ) {
        skip = true;
      }
    }
    // only look for more if this one isn't recorded
    if (!skip) {
      const currentIdString =
        PassedCourseList[i].subject + "-" + PassedCourseList[i].number;
      const tempCatArr: number[] = [];
      for (let j = i; j < PassedCourseList.length; j++) {
        if (
          currentIdString ===
          PassedCourseList[j].subject + "-" + PassedCourseList[j].number
        ) {
          tempCatArr.push(PassedCourseList[j].idCategory);
        }
      }
      if (tempCatArr.length > 1) {
        tempArr.push({ idString: currentIdString, categories: tempCatArr });
      }
    }
  }
  return tempArr;
}
