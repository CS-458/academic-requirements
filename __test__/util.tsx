import { jest } from "@jest/globals";
import type { MatcherFunction } from "expect";
import { DndProvider } from "react-dnd";
import { HTML5Backend } from "react-dnd-html5-backend";
import { QueryClient, QueryClientProvider, UseQueryResult } from "react-query";
import { testApiHandler } from "next-test-api-route-handler";
import { NextApiHandler } from "next";
import {
  Matcher,
  screen,
  render as testRender,
  RenderResult,
  waitFor
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { UserEvent } from "@testing-library/user-event/dist/types/setup/setup";
import { User, UserMajor } from "../services/user";
import { VerifyIdTokenOptions, LoginTicket } from "google-auth-library";
import { updateClient } from "../services/login";
import { PromisedDatabase } from "promised-sqlite3";
import { setUserDb } from "../services/sql";

import "../pages/api";
import { dragAndDrop } from "./dragDrop";

interface UserExt {
  /// Select an element from an `AutoComplete` dropdown
  selectAutocomplete: typeof selectAutocomplete;
  /// Drag an element from one place to another
  drag: typeof dragAndDrop;
}

async function selectAutocomplete(
  this: UserEvent,
  label: Matcher,
  option: Matcher
): Promise<void> {
  await this.click(screen.getByTestId("test-root-element"));
  const input = screen.getByLabelText(label, { selector: "input" });
  await waitFor(async () => {
    expect(input).not.toBeDisabled();
  });
  await this.click(input);
  try {
    await this.click(
      await screen.findByText(option, undefined, { timeout: 10000 })
    );
  } catch (e) {
    const rootEl = screen.getByTestId("test-root-element");
    const popper = rootEl?.querySelector(".MuiAutocomplete-popper");
    const html = popper?.innerHTML;
    throw new Error(
      `Option ${option} was not found in dropdown.\nOptions: ${html}`
    );
  }
}

export function parentEl(el: HTMLElement, name: string): HTMLElement {
  while (!el.classList.contains(`${name}-root`)) {
    const next = el.parentElement;
    if (next == null) {
      throw new Error(`Element does not have a wrapper with name ${name}`);
    }
    el = next;
  }
  return el;
}

/// userEvent.setup(), but with some additional convience methods
export function setupUser(): UserEvent & UserExt {
  return {
    ...userEvent.setup(),
    selectAutocomplete,
    drag: dragAndDrop
  };
}

// const modules: { [key: string]: NextApiHandler | Promise<NextApiHandler> } = {};
// declare function fetch2(input: RequestInfo | URL, init?: RequestInit): Promise<Response>;
/// Executes a request against an Api Route, roughly equavelent with `fetch`.
export async function fetchApiRoute(
  url: RequestInfo | URL,
  init?: RequestInit
): Promise<Response> {
  const res = { resolve: (_a: any) => { }, reject: (_e: any) => { } };
  const result: Promise<Response> = new Promise((resolve, reject) => {
    res.resolve = resolve;
    res.reject = reject;
  });

  let fullUrl: string | undefined;
  let path;
  let params;
  if (url instanceof URL) {
    fullUrl = url.href;
    path = url.pathname;
    params = url.search;
  } else if (typeof url === "string") {
    fullUrl = url;
    [path, params] = url.split("?");
  } else {
    fullUrl = url.url;
    [path, params] = url.url.split("?");
  }

  const handler: NextApiHandler = await import(`../pages/${path}`).catch(
    (e) => {
      if (jest.isEnvironmentTornDown()) {
        console.log("Request made after environment torn down");
      } else {
        console.error(e);
      }
    }
  );
  await testApiHandler({
    url: fullUrl,
    handler,
    params: params?.split("&").reduce((a, b) => {
      const [name, value] = b.split("=");
      return { ...a, [name]: value };
    }, {}),
    test: async ({ fetch }) => {
      const inner = await fetch(init);
      res.resolve(inner);
    }
  });
  return await result;
}

/// Calls fetchApiRoute, and decodes the response as JSON
export async function fetchApiJson(
  url: RequestInfo | URL,
  init?: RequestInit
): Promise<any> {
  return await (await fetchApiRoute(url, init)).json();
}

/**
 *  Wrap Elements with QueryClient & DndProvider similar to _app.tsx
 *
 *  Also overides window.fetch to use fetchApiRoute for testing
 * */
export function render(children: JSX.Element | JSX.Element[]): RenderResult {
  window.fetch = fetchApiRoute;
  return testRender(
    <div data-testid="test-root-element">
      <QueryClientProvider client={new QueryClient()}>
        <DndProvider backend={HTML5Backend}>{children}</DndProvider>
      </QueryClientProvider>
    </div>
  );
}

const localStorage: { [key: string]: string | null } = {};

export function getMockStorage(key: string): string | null {
  return localStorage[key];
}

export function setMockStorage(key: string, value: string): void {
  localStorage[key] = value;
}

export function removeMockStorage(key: string): void {
  localStorage[key] = null;
}

export function clearMockStorage(): void {
  for (const k in localStorage) {
    localStorage[k] = null;
  }
}

/// Configure local storage and set pre-existing data in local storage
export function buildLocalStorage(user?: UserMajor): void {
  window.localStorage.getItem = getMockStorage;
  window.localStorage.setItem = setMockStorage;
  window.localStorage.removeItem = removeMockStorage;
  window.localStorage.clear = clearMockStorage;
  if (user !== undefined) {
    window.localStorage.setItem("user_major", JSON.stringify(user));
  }
}

export function mockUseQuery<T>(
  _key: any,
  fetch: () => Promise<T>
): UseQueryResult<T> & { fetch: Promise<T> } {
  return {
    fetch: fetch(),
    data: undefined,
    error: null,
    isError: false,
    isIdle: false,
    isLoading: true,
    isLoadingError: false,
    isRefetchError: false,
    isSuccess: false,
    status: "loading",
    dataUpdatedAt: Date.now(),
    errorUpdatedAt: Date.now(),
    failureCount: 0,
    errorUpdateCount: 0,
    isFetched: true,
    isFetchedAfterMount: true,
    isFetching: false,
    isPlaceholderData: false,
    isPreviousData: false,
    isRefetching: false,
    isStale: false,
    refetch: async (_options: any) => {
      throw new Error("");
    },
    remove: () => { }
  };
}

export function setupStandaloneQuery(): void {
  jest.mock("react-query", () => ({
    useQuery: mockUseQuery
  }));
}

export async function waitForQuery<T>(
  query: () => UseQueryResult<T>
): Promise<T> {
  // @ts-expect-error fetch only exists on our mock
  const temp = await query().fetch;
  return temp;
}

const toBeUnique: MatcherFunction<[]> = (actual) => {
  if (!Array.isArray(actual)) {
    throw new Error("actual is not an object");
  }
  if (actual === null) {
    return {
      message: () => "null is not an array",
      pass: false
    };
  }
  for (const i of actual) {
    const pos = actual.find((val, idx) => idx > i && actual[i] === val);
    if (pos !== undefined) {
      return {
        message: () => `Indicies ${i}, and ${pos} are the same`,
        pass: false
      };
    }
  }
  return {
    message: () => "All elements are unique",
    pass: true
  };
};

const toBeHidden: MatcherFunction<[]> = (actual) => {
  if (!(actual instanceof HTMLElement)) {
    throw new Error("actual is not an HTMLElement");
  }
  let cur = actual;
  console.log("child", cur.ariaHidden);
  while (cur.ariaHidden == null || cur.ariaHidden === "false") {
    if (cur.dataset.testid === "Semester") {
      console.log(cur.ariaHidden);
    }
    const next = cur.parentElement;
    if (next === null) {
      return {
        message: () => "Element (or parent) is not aria-hidden",
        pass: false
      };
    }
    cur = next;
  }
  return {
    message: () => "Element (or parent) is aria-hidden",
    pass: true
  };
};

expect.extend({
  toBeUnique,
  toBeHidden
});

declare module "expect" {
  interface AsymmetricMatchers {
    toBeUnique: () => void;
    toBeHidden: () => void;
  }
  interface Matchers<R> {
    toBeUnique: () => R;
    toBeHidden: () => R;
  }
}

export function courseSemestersCheck(semesters: string): void {
  /*
    Available options: FA,WI,SP,SU

    semesters can start with any of the options
    then be followed by a comma and any of the options

    Note: Data such as FA,FA will match but FA,bad_data will not
  */
  // Expect the format to match
  expect(semesters).toMatch(/^(FA|WI|SP|SU)(,(FA|WI|SP|SU))*$/);
}

export function createMockToken(): void {
  updateClient((c) => {
    c.verifyIdToken = async (options: VerifyIdTokenOptions) => {
      const [test, id] = options.idToken.split(":");
      if (test !== "TEST_TOKEN") {
        throw new Error("Invalid Token");
      }
      return new LoginTicket("", {
        sub: id,
        aud: c._clientId ?? "",
        exp: Date.now() + 1000000,
        iat: Date.now() - 100,
        iss: "https://accounts.google.com",
        at_hash: ""
      });
    };
  });
}

export function mockToken(id: string): string {
  return `TEST_TOKEN:${id}`;
}

export function mockUserInfo(id: string): User {
  return {
    cred: mockToken(id),
    info: {
      sub: id,
      picture: "",
      name: id,
      email: `${id}@gmail.com`
    }
  };
}

const db = {
  db: new PromisedDatabase(),
  opened: false
};

export async function setupMockUserDB(): Promise<void> {
  createMockToken();
  setUserDb(db.db);
  if (!db.opened) {
    db.opened = true;
    await db.db.open(":memory:");
    await db.db.run(
      "CREATE TABLE `user` ( `idUser` TEXT PRIMARY KEY, `role` TEXT NOT NULL)"
    );
    await db.db.run(
      "CREATE TABLE `schedule` ( `userID` TEXT, `name` TEXT, `timestamp` DATE DEFAULT (datetime('now','localtime')), `scheduleData` TEXT NOT NULL, PRIMARY KEY(`userID`, `name`))"
    );
  }
}
