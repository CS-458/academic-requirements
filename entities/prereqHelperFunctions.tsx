import { SemesterType, CourseType } from "./four_year_plan";
import { userMajor } from "../services/user";
import StringProcessing from "./StringProcessing";
//  Returns if a course is already in a semester's index
export function courseAlreadyInSemester(course: CourseType, semesterIndex: number, semesters: SemesterType[]): boolean {
  if (semesterIndex > -1 && semesterIndex < semesters.length) {
    semesters[semesterIndex].courses.forEach((x: any) => {
      if (x === course) {
        return true;
      }
    });
  }
  return false;
}

//  Get all courses in previous semesters
//  param semesterIndex -> current semester index
function getPreviousSemesterCourses(semesterIndex: number, semesters: SemesterType[]): Array<string> {
  const previousCourses = new Array<string>();
  if (semesterIndex > -1 && semesterIndex < semesters.length) {
    semesters.forEach((currSemester) => {
      if (currSemester.semesterNumber - 1 < semesterIndex) {
        currSemester.courses.forEach((x: any) => {
          previousCourses.push(x.subject + "-" + x.number);
        });
      }
    });
  }

  //  Append completed courses to the array
  userMajor()?.completed_courses.forEach((x: string) => {
    previousCourses.push(x);
  });

  return previousCourses;
}

//  Get all Course objects in current semester
//  param semesterIndex -> current semester index
function getSemesterCourses(semesterIndex: number, semesters: SemesterType[]): Array<CourseType> {
  const semCourses = new Array<CourseType>();
  if (semesterIndex > -1 && semesterIndex < semesters.length) {
    semesters[semesterIndex].courses.forEach((x: any) => {
      semCourses.push(x);
    });
  }
  return semCourses;
}

//  Get all courses (string) in current semester
//  param semesterIndex -> current semester index
export function getSemesterCoursesNames(semesterIndex: number, semesters: SemesterType[]): Array<string> {
  const semCourses = new Array<string>();
  if (semesterIndex > -1 && semesterIndex < semesters.length) {
    semesters[semesterIndex].courses.forEach((x: CourseType) => {
      semCourses.push(x.subject + "-" + x.number);
    });
  }
  return semCourses;
}

//  This function checks if every course passes the prerequisite check when moving a course
//  out of a semester
export function preReqCheckAllCoursesPastSemester(
  courseToRemove: CourseType,
  courseSemesterIndex: number,
  showMessage: boolean,
  movedRight: boolean,
  draggedOut: boolean,
  semesters: SemesterType[],
  warningPrerequisiteCourses: CourseType[]
): { vis: boolean, error: string, warning: CourseType[] } {
  // prereqCheck will be used to check prerequisites
  const preReqCheck = new StringProcessing();

  let visibility = false;
  let errorMessage = "";
  let setWarning: CourseType[] = [];

  //  Get the course names in the previous semesters
  const previousCourses = getPreviousSemesterCourses(courseSemesterIndex === -1 ? 0 : courseSemesterIndex, semesters);

  //  Get the current courses in the current semester
  let currentCourses = getSemesterCourses(courseSemesterIndex === -1 ? 0 : courseSemesterIndex, semesters);
  let currentCoursesNames = getSemesterCoursesNames(courseSemesterIndex === -1 ? 0 : courseSemesterIndex, semesters);

  const failedCoursesList = new Array<CourseType>();
  const failedCoursesNoWarning = new Array<CourseType>();

  semesters.forEach((currSemester, index) => {
    if (currSemester.semesterNumber - 1 >= courseSemesterIndex) {
      //  Check every course in the current semester passes the prerequisites and push any failed
      //  prerequisites to the failedCoursesList
      currentCourses.forEach((x) => {
        if (
          !preReqCheck.courseInListCheck(
            x !== undefined ? x.preReq : "",
            previousCourses,
            currentCoursesNames
          ).returnValue
        ) {
          failedCoursesList.push(x);
        }
        //  If the course prereq fails, but not due to moving the course,
        //  add it to the failedCoursesNoWarning list
        console.log("previous", previousCourses);
        console.log("current", currentCourses);
        if (
          !preReqCheck.courseInListCheck(
            x !== undefined ? x.preReq : "",
            previousCourses,
            currentCoursesNames
          )
            .failedString.includes(courseToRemove.subject + "-" + courseToRemove.number)
        ) {
          console.log("setting it");
          failedCoursesNoWarning.push(x);
        }
      });

      //  Append the current semester to the previous courses semester
      currentCoursesNames.forEach((x) => {
        previousCourses.push(x);
      });

      //  Update the current course lists to be for the next semester
      if (index + 1 < semesters.length && semesters[index + 1].courses !== undefined) {
        currentCourses = getSemesterCourses(index + 1, semesters);
        currentCoursesNames = getSemesterCoursesNames(index + 1, semesters);
      }
    }
  });

  //  Prepping variables for modifying warningPrerequisitesCourses
  let found = false;
  let tempWarningCourses = warningPrerequisiteCourses;
  const initialPreviousCourses = new Array<CourseType>();

  //  Add previous courses to initialPreviousCourses (the course object, not the strings)
  semesters.forEach((x, index) => {
    if (index < courseSemesterIndex) {
      x.courses.forEach((y: any) => {
        initialPreviousCourses.push(y);
      });
    }
  });

  // Remove any courses that were marked as warning, but now have resolved prerequisites
  if (!movedRight) {
    warningPrerequisiteCourses.forEach((currentWarningCourse) => {
      if (initialPreviousCourses.find((prevCourse) => prevCourse === currentWarningCourse) === undefined) {
        failedCoursesList.forEach((currentFailedCourse) => {
          if (currentWarningCourse === currentFailedCourse) {
            found = true;
          }
        });

        //  If the currently selected course in the warningCourses now passes the prerequisites
        if (!found) {
          const temp = new Array<CourseType>();
          //  Replace warningCourses with all courses but the currently selected warningCourse
          tempWarningCourses.forEach((temporaryCurrentWarningCourse) => {
            //  Carry on if the tempWarningCourse is not in a previous semester
            if (temporaryCurrentWarningCourse !== currentWarningCourse) {
              temp.push(temporaryCurrentWarningCourse);
            }
          });
          tempWarningCourses = temp;
        }
        found = false;
      }
    });

    //  Update the warning courses to remove the currently now-satisifed prereqs course
    setWarning = tempWarningCourses;
  }

  //  If any courses have failed, notify the user of each course that failed
  if (showMessage && failedCoursesList.length > 0) {
    console.log("creating message");
    let message = "";
    //  Push each failed course to the warningCourses
    console.log("list", failedCoursesList);
    failedCoursesList.forEach((x) => {
      if (warningPrerequisiteCourses.find((z) => z === x) === undefined) {
        const temp = warningPrerequisiteCourses;
        temp.push(x);
        setWarning = temp;
      }
      //  If the course is failing, but not due to the latest course move, modify the warning message
      console.log(failedCoursesNoWarning);
      console.log(x);
      if (failedCoursesNoWarning.find((z) => z === x) === undefined) {
        console.log("writing");
        message.length > 0 ? (message = message + "," + x.subject + "-" + x.number) : (message = message + x.subject + "-" + x.number);
      }
    });
    //  Show a warning stating that the classes failed the prereqs
    console.log(!message.includes(courseToRemove.subject + "-" + courseToRemove.number));
    console.log(message.length);
    if (!message.includes(courseToRemove.subject + "-" + courseToRemove.number) && message.length > 0) {
      console.log("here");
      visibility = true;
      errorMessage = "WARNING! " + courseToRemove.subject + "-" + courseToRemove.number +
        " is a prerequisite for the following courses: " + message;
    }
  }

  return { vis: visibility, error: errorMessage, warning: setWarning };
}
