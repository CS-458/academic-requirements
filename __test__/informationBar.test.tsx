import "@testing-library/jest-dom";
import { screen } from "@testing-library/react";
import { setupUser, render } from "./util";
import InformationDrawer from "../components/InformationBar";
import { RequirementComponentType } from "../entities/four_year_plan";

const requirementsDisplayList: RequirementComponentType[] = [{
  courseCount: null,
  courseCountTaken: 0,
  courseReqs: "CNIT-133,CNIT-134,CNIT-301,CNIT-361,CNIT-382,CNIT-383,CNIT-484,CS-458,CS-480",
  coursesTaken: "",
  creditCount: 32,
  creditCountTaken: 0,
  idCategory: 19,
  name: "CS - Cyber Security and Secure Software Development",
  parentCategory: null,
  percentage: 0,
  shortName: "CS CSSSD",
  inheritedCredits: 0
}, {
  courseCount: null,
  courseCountTaken: 0,
  courseReqs: null,
  coursesTaken: "",
  creditCount: 6,
  creditCountTaken: 0,
  idCategory: 22,
  name: "Global Perspectives (GLP)",
  parentCategory: null,
  percentage: 0,
  shortName: "GLP",
  inheritedCredits: 0
}];

test("Render the Information Bar on the Schedule Page", async () => {
  const user = setupUser();
  const bar = render(<InformationDrawer requirementsDisplay={requirementsDisplayList} />);
  expect(bar.baseElement).toMatchSnapshot();
  expect(screen.getByText(/Major/i)).toBeInTheDocument();
  expect(screen.getByText(/CS CSSSD/i)).toBeInTheDocument();
  expect(screen.getByText(/Gen Eds/i)).toBeInTheDocument();
  expect(screen.getByText(/GLP/i)).toBeInTheDocument();
  try {
    screen.getByText(/Global Perspectives (GLP)/i);
    expect(false);
  } catch (error) {
    expect(true);
  }
  const openButton = screen.getByTestId("openDrawer");
  await user.click(openButton);
  expect(screen.getByText(/Global Perspectives (GLP)/i)).toBeInTheDocument();
});
