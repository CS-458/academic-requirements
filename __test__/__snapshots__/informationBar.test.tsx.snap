// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the Information Bar on the Schedule Page 1`] = `
<body>
  <div>
    <div
      data-testid="test-root-element"
    >
      <div
        class="MuiBox-root css-k008qs"
      >
        <div
          class="MuiDrawer-root MuiDrawer-docked css-1txhcbl-MuiDrawer-docked"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorRight MuiDrawer-paperAnchorDockedRight css-18sg6k4-MuiPaper-root-MuiDrawer-paper"
          >
            <div
              class="css-15lei78"
            />
            <div
              class="css-ak1ss8"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                data-testid="openDrawer"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ChevronLeftIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              aria-labelledby="simple-tab-0"
              id="simple-tabpanel-0"
              role="tabpanel"
            >
              <div
                class="MuiBox-root css-hpgf8j"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
                >
                  Major
                </p>
                <hr
                  class="MuiDivider-root MuiDivider-fullWidth css-a5zps3-MuiDivider-root"
                />
                <div>
                  <div
                    class="RequirementText"
                    data-testid="requirement"
                  >
                    <div
                      class="requirementName"
                    >
                      CS CSSSD
                    </div>
                    <div
                      class="percentage"
                    >
                      0%
                    </div>
                  </div>
                </div>
                <div>
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
                  >
                    Gen Eds
                  </p>
                  <hr
                    class="MuiDivider-root MuiDivider-fullWidth css-a5zps3-MuiDivider-root"
                  />
                  <div
                    class="RequirementText"
                    data-testid="requirement"
                  >
                    <div
                      class="requirementName"
                    >
                      GLP
                    </div>
                    <div
                      class="percentage"
                    >
                      0%
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
